csrc/dynamic_transform.cpp

```diff
@@ -385,6 +648,219 @@ void DynamicTransformConcretizer::concretize() {
   }
 }
 
+void DynamicTransformConcretizer::removeEmptyBranches() {
+  auto fusion = FusionGuard::getCurFusion();
+  for (const auto& empty_tv_descr : info_->getEmptyTensors()) {
+    auto tv = info_->initialInfo()->lookUpTV(empty_tv_descr.tv_name);
+    auto rfactor = TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+    std::vector<Val*> new_shape;
+    new_shape.reserve(rfactor.size());
+    for (auto id : rfactor) {
+      new_shape.push_back(id->getMaybeExpandedExtent());
+    }
+    for (auto ax : empty_tv_descr.empty_axes) {
+      // Hard-code zero extent for empty axes. This lets us detect empty input
+      // and output tensors during scheduling/execution.
+      registerConcretization(new_shape[ax], fusion->zeroVal());
+      new_shape[ax] = fusion->zeroVal();
+    }
+
+    auto hasEmptyRootReductionAxis = [&empty_tv_descr](TensorView* out_tv) {
+      return std::any_of(
+          empty_tv_descr.empty_axes.begin(),
+          empty_tv_descr.empty_axes.end(),
+          [&out_tv](size_t ax) {
+            return out_tv->getRootDomain().at(ax)->isReduction();
+          });
+    };
+
+    // Given a TensorView get a vector of its maybeRFactor maybeExpandedExtents
+    auto orig_shape = [](TensorView* out_tv) -> std::vector<Val*> {
+      const auto& rfactor =
+          TensorDomain::noReductions(out_tv->getMaybeRFactorDomain());
+      std::vector<Val*> out_shape;
+      out_shape.reserve(rfactor.size());
+      for (const auto id : rfactor) {
+        out_shape.push_back(id->getMaybeExpandedExtent());
```

Comment:
Why is a domain expanded? Can we create a tensor with an expanded domain with a full op?

Reply:
This lambda is applied to outputs from reductions, Welfords, and pads. If those have inputs with non-reduced and non-padded dimensions that are expanded, they'll be preserved as expanded in the output. We could instead replace with a full followed by an expand, but I figured we could just go ahead and realize the expand on an empty tensor instead for simplicity.